tools/linters/ll.go

```diff
@@ -0,0 +1,219 @@
+	"github.com/golangci/golangci-lint/pkg/lint/linter"
+	"github.com/golangci/golangci-lint/pkg/result"
+	"golang.org/x/tools/go/analysis"
+)
+
+const (
+	linterName               = "lll"
+	goCommentDirectivePrefix = "//go:"
+)
+
+// New creates a new lll linter from the given settings. It satisfies the
+// signature required by the golangci-lint linter for plugins.
+func New(settings *config.LllSettings) *goanalysis.Linter {
+	var (
+		mu        sync.Mutex
+		resIssues []goanalysis.Issue
+	)
+
+	analyzer := &analysis.Analyzer{
+		Name: linterName,
+		Doc:  goanalysis.TheOnlyanalyzerDoc,
+		Run: func(pass *analysis.Pass) (any, error) {
+			issues, err := runLll(pass, settings)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(issues) == 0 {
+				return nil, nil
+			}
+
+			mu.Lock()
+			resIssues = append(resIssues, issues...)
+			mu.Unlock()
+
+			return nil, nil
+		},
+	}
+
+	return goanalysis.NewLinter(
+		linterName, "Reports long lines",
+		[]*analysis.Analyzer{analyzer}, nil,
+	).WithIssuesReporter(func(*linter.Context) []goanalysis.Issue {
+		return resIssues
+	}).WithLoadMode(goanalysis.LoadModeSyntax)
+}
+
+func runLll(pass *analysis.Pass, settings *config.LllSettings) (
+	[]goanalysis.Issue, error) {
+
+	var (
+		fileNames = getFileNames(pass)
+		spaces    = strings.Repeat(" ", settings.TabWidth)
+		issues    []goanalysis.Issue
+	)
+
+	for _, f := range fileNames {
+		lintIssues, err := getLLLIssuesForFile(
+			f, settings.LineLength, spaces,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		for i := range lintIssues {
+			issues = append(issues, goanalysis.NewIssue(
+				&lintIssues[i], pass,
+			))
+		}
+	}
+
+	return issues, nil
+}
+
+func getLLLIssuesForFile(filename string, maxLineLen int,
+	tabSpaces string) ([]result.Issue, error) {
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("can't open file %s: %w", filename, err)
+	}
+	defer f.Close()
+
+	var (
+		res                []result.Issue
+		lineNumber         int
+		multiImportEnabled bool
+	)
+
+	// Scan over each line.
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		lineNumber++
+
+		// Replace all tabs with spaces.
+		line := scanner.Text()
+		line = strings.ReplaceAll(line, "\t", tabSpaces)
+
+		// Ignore any //go: directives since these cant be wrapped onto
+		// a new line.
+		if strings.HasPrefix(line, goCommentDirectivePrefix) {
+			continue
+		}
+
+		// We never want the linter to run on imports since these cannot
+		// be wrapped onto a new line. If this is a single line import
+		// we can skip the line entirely. If this is a multi-line import
+		// skip until the closing bracket.
+		//
+		// NOTE: We trim the line space around the line here purely for
+		// the purpose of being able to test this part of the linter
+		// without the risk of the `gosimports` tool reformatting the
+		// test case and removing the import.
+		if strings.HasPrefix(strings.TrimSpace(line), "import") {
```

Comment:
here is the only change to the OG code: it is done purely so we can write tests that dont get edited by `gosimports` 😂